server/src/main/java/org/opensearch/gateway/ShardBatchCache.java

```diff
@@ -0,0 +1,241 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Cache implementation of transport actions returning batch of shards data in the response. Cache uses a specific
+ * NodeEntry class that stores the data in array format. To keep the class generic for primary or replica, all
+ * functions are stored during object creation.
+ *
+ * @param <T> Response type of transport action.
+ * @param <V> Data type of shard level response.
+ */
+public class ShardBatchCache<T extends BaseNodeResponse, V extends BaseShardResponse> extends BaseShardCache<T> {
+    private final Map<String, NodeEntry<V>> cache;
+    private final Map<ShardId, Integer> shardIdToArray; // used for mapping array index for a shard
+    private final AtomicInteger shardIdIndex;
+    private final int batchSize;
+    private final Class<V> shardResponseClass;
+    private final BiFunction<DiscoveryNode, Map<ShardId, V>, T> responseConstructor;
+    private final Map<Integer, ShardId> arrayToShardId;
+    private final Function<T, Map<ShardId, V>> shardsBatchDataGetter;
+    private final Supplier<V> emptyResponseBuilder;
+    private final Consumer<ShardId> handleFailedShard;
+
+    public ShardBatchCache(
+        Logger logger,
+        String type,
+        Map<ShardId, ShardAttributes> shardToCustomDataPath,
+        String logKey,
+        Class<V> clazz,
+        BiFunction<DiscoveryNode, Map<ShardId, V>, T> responseConstructor,
+        Function<T, Map<ShardId, V>> shardsBatchDataGetter,
+        Supplier<V> emptyResponseBuilder,
+        Consumer<ShardId> handleFailedShard
+    ) {
+        super(logger, logKey, type);
+        this.batchSize = shardToCustomDataPath.size();
+        fillShardIdKeys(shardToCustomDataPath.keySet());
+        this.shardResponseClass = clazz;
+        this.responseConstructor = responseConstructor;
+        this.shardsBatchDataGetter = shardsBatchDataGetter;
+        this.emptyResponseBuilder = emptyResponseBuilder;
+        cache = new HashMap<>();
+        shardIdToArray = new HashMap<>();
+        arrayToShardId = new HashMap<>();
+        shardIdIndex = new AtomicInteger();
+        this.handleFailedShard = handleFailedShard;
+    }
+
+    @Override
+    public Map<String, ? extends BaseNodeEntry> getCache() {
+        return cache;
+    }
+
+    @Override
+    public void deleteData(ShardId shardId) {
+        if (shardIdToArray.containsKey(shardId)) {
+            Integer shardIdIndex = shardIdToArray.remove(shardId);
+            for (String nodeId : cache.keySet()) {
+                cache.get(nodeId).clearShard(shardIdIndex);
+            }
+        }
+    }
+
+    @Override
+    public Map<DiscoveryNode, T> getCacheData(DiscoveryNodes nodes, Set<String> failedNodes) {
+        refreshReverseIdMap();
```

Comment:
why is refresh required given that the shardId is not changing with get ?

Reply:
True. ShardId and index mapping is not changing. 
it's not actually a refresh, my agenda was to not fill up this map during init. 
As we only need this just before we need to give full cache's data (i.e. getCacheData), I thought it would be useless to fill this map during fetch. 

Fetch from all nodes can take time and this map is only needed when we need to return the data, after all fetching is done.  
@rajiv-kv  Thoughts ?

Reply:
Makes sense; Can we rename only the lines of `loadCache` instead of `refresh` ? Also, does this needs to be refreshed for every get.

Reply:
removed the reverse map.